### ISSUE TITLE
A general model for update in layer

### ISSUE THREAD

**Issue by YUcxovo** [linsyking/Messenger#13]
the update function of layers with components or game components can be modelled in a general way.
The update function can be divided into 4 parts: 
- update basic data (remove dead game components, etc.) 
- update component list by themselves
- decide the msg that need to be sent and distribute them (collisions, etc.) 
- handle all the component msg from the former steps

A `distributor` type : `Env /*alias of Env C in Layer*/ -> Model -> list (ComponentTarget, ComponentMsg)` is designed for part 3. Specifically, is will be implement by using a `distributor` and `upateobjectswithtarget` func
Finally, the update function will be: `updatebasic  -> updatecomponents -> updatecomponentsbydistributor -> handlecomponentmsgs`, and for users, they only need to deal with updatebasic, distributors and handler.
*Basically it is not recommend for users to put both `GameComponent` and `Component` in one layer.*

**Comment by linsyking** [comment]
Great! BTW, we have changed `EnvC` to `Env c` and `Env` to `Env ()`.

**Comment by linsyking** [comment]
Could you illustrate more on how users will compose a layer under your general model?

**Comment by YUcxovo** [comment]
> Could you illustrate more on how users will compose a layer under your general model?

Let's take a layer with component for example. My model just affect the update part in fact.
the `updateModel` function is no need to be modified by users. It will be divided into four parts:
- `updateBasic : Env C -> Model -> ( Model, List ( LayerTarget, LayerMsg ), Env )`
- `updateComponents`  *already exists in `Env.Component.ComponentHandler`*
- `distributorComponents : Env C -> Model -> ( Model, ( List ( LayerTarget, LayerMsg ), List (ComponentTarget, ComponentMsg)), Env )`  *The output can be simplified, but I'm tangled in deleting which part of it*
Then the result of `List (ComponentTarget, ComponentMsg)` will be used in `updateComponentswithTarget`
- `handleComponentMsg : Env -> ComponentMsg -> Model -> ( Model, List ( LayerTarget, LayerMsg ), Env )`

**Users will only need to do their work in `updateBasic` func, `distributorComponents` func and `handleComponentMsg` func.**
